scipy/spatial/ckdtree/src/find_min_max.cxx

```diff
@@ -0,0 +1,58 @@
+#include <vector>
+#include <algorithm>
+#include <numeric>
+#include "ckdtree_decl.h"
+
+
+bool
+compare_idx(ckdtree *self, int idx1, int idx2, const int dimension, const bool max){
+    if(max)
+        std::swap(idx1, idx2);
+    return self->raw_data[(self->raw_indices[idx1])*(self->m) + dimension] <
+           self->raw_data[(self->raw_indices[idx2])*(self->m) + dimension];
+}
+
+std::vector<ckdtree_intp_t>
+find_min_max(ckdtree *self, ckdtreenode *node, const int dimension, const int n, const bool max){
+    if(node->split_dim == dimension)
+        return find_min_max(self, max ? node->greater : node->less, dimension, n, max);
+
+    else if(node->split_dim == -1){
+        std::vector<ckdtree_intp_t> indices(node->children);
+        std::iota(indices.begin(), indices.end(), node->start_idx);
+
+        std::sort(indices.begin(), indices.end(),
+            [self, dimension, max](ckdtree_intp_t idx1, ckdtree_intp_t idx2){
+                return compare_idx(self, idx1, idx2, dimension, max);
+            });
+
+        indices.resize(n);
+        return indices;
+    }
+    else{
+        std::vector<ckdtree_intp_t> indices = find_min_max(self, node->less, dimension, n, max);
+        std::vector<ckdtree_intp_t> to_append = find_min_max(self, node->greater, dimension, n, max);
```

Comment:
I'm sorry, I just realised I was completely wrong when I said the bounds could be recomputed in O(log n). I was thinking of a similar data structure that stores the mins/maxs for all the tree nodes, not just the root. With that data structure, recomputing the bounds is much cheaper and also much more important to query performance.

What you've written here is something weird like O(n * 2**-m) for a well balanced tree, or O(n) worst case. I'd be interested to know how this compares to a simple loop over `raw_data` but I now suspect it's not worth it.

Reply:
I've tried to compare `np.amin(dataset[:, dim])` with `dataset[ckdtree.find_min(dim), dim]` and it is convenient only if `self.m` is  very small (as you have highlighted). In other cases the two times are about the same. 

However, a C++ implementation of _find_min_ and _find_max_ is needed if we want to perform a rebalancing, I was thinking about that:
-Implement multiple remove and multiple add
-During removal and addition, if a children node has _many_ more points of the other, we remove some points from the heaviest and we add to the other. To find points to remove we need find_min/find_max.

If you think that a simple loop over raw_data could be better, I can try to implement it.

Reply:
The problem for me is that that you need to call this function 2m times to recalculate the entire hyper-rectangle bounds. All things considered, I think you were right in the first place and this probably isn't worth doing.

As for rebalancing, you need to be very careful. At some point it becomes faster and simpler to just build a new tree from scratch.

Reply:
Ok, well, I can revert to "adjust size update" commit without problems